atp/client.go

```diff
@@ -199,23 +203,20 @@ func (c *client) Execute(
 	}
 	c.logger.Debugf("Step '%s' started, waiting for response...", stepData.ID)
 
-	defer c.handleStepComplete(stepData.RunID, receivedSignals)
+	if receivedSignals != nil {
+		defer c.handleStepComplete(stepData.RunID)
+	}
```

Comment:
So this change seems to assume (without any proof I can see) that `receivedSignals` is linked to `c.runningSignalReceiveLoops[runID]`. Is that really valid/safe? Given the behavior of the function, it seems that the minor optimization of dropping the `defer` may not be worthwhile ...

Reply:
The test at line 206 here corresponds to the test/code just above it at line 187:  the call to `c.executeWriteLoop()` is what puts the value from `receivedSignals` into the entry in `runningSignalReceiveLoops`; the code here at line 207 is (one of the places...) that removes it again, so we only need to make the call here if we made the call above.

Reply:
OK; so 187 is *also* dependent on `c.atpVersion > 1`, so should this defer be pushed up into that block, maybe immediately after scheduling the execute thread?

Reply:
Yeah.  That seems like a _really_ good idea.  (Like -- duh! -- I should have thought of that!)

Reply:
Um, now that I look at it....  There doesn't seem to be a real need for `handleStepComplete()` and, in fact, there's no need even for `runningSignalReceiveLoops`.  So, in my latest revision, I've taken the liberty of removing them both entirely.

@jaredoconnell, am I nuts?